axelor-business-project/src/main/java/com/axelor/apps/businessproject/web/ProjectTaskController.java

```diff
@@ -137,4 +143,57 @@ public void getProjectTaskFinancialReportingData(ActionRequest request, ActionRe
     data.put("forecastMarkup", request.getData().get("forecastMarkup"));
     response.setData(List.of(data));
   }
+
+  public void generatePurchaseOrder(ActionRequest request, ActionResponse response) {
+    ProjectTask projectTask = null;
+    if (request.getContext().get("_projectTaskId") != null) {
+      projectTask =
+          Beans.get(ProjectTaskRepository.class)
+              .find(Long.valueOf((Integer) request.getContext().get("_projectTaskId")));
+    } else {
+      projectTask = request.getContext().asType(ProjectTask.class);
+    }
+    SaleOrderLine saleOrderLine = projectTask.getSaleOrderLine();
+    Partner supplierPartner = null;
+    String saleOrderLinesIdStr = null;
+
+    if (request.getContext().get("supplierPartnerSelect") != null) {
+      supplierPartner =
+          JPA.em()
+              .find(
+                  Partner.class,
+                  Long.valueOf(
+                      (Integer)
+                          ((Map) request.getContext().get("supplierPartnerSelect")).get("id")));
+      saleOrderLinesIdStr = (String) request.getContext().get("saleOrderLineIdSelected");
+    }
+    Map<String, Object> view = null;
+
+    try {
+      if (saleOrderLine != null) {
+        SaleOrder saleOrder = saleOrderLine.getSaleOrder();
+        List<SaleOrderLine> saleOrderLines = List.of(saleOrderLine);
+        view =
+            Beans.get(PurchaseOrderFromSaleOrderLinesService.class)
+                .generatePurchaseOrdersFromSOLines(
+                    saleOrder, saleOrderLines, supplierPartner, saleOrderLinesIdStr);
+      } else {
+        view =
+            Beans.get(PurchaseOrderProjectService.class)
+                .generateEmptyPurchaseOrderFromProjectTask(projectTask, supplierPartner);
+      }
+    } catch (Exception e) {
+      TraceBackService.trace(response, e);
+    }
```

Comment:
What is this exception for ? You should throw an AxelorException to show the error in the view (if not, it will only generate a TraceBack, without alerting the user)